src/server/frontend_wayland/xdg_decoration_unstable_v1.cpp

```diff
@@ -60,60 +62,68 @@ class ToplevelsWithDecorations
 class XdgDecorationManagerV1 : public wayland::XdgDecorationManagerV1
 {
 public:
-    XdgDecorationManagerV1(wl_resource* resource);
+    XdgDecorationManagerV1(wl_resource* resource, std::shared_ptr<DecorationStrategy> strategy);
 
     class Global : public wayland::XdgDecorationManagerV1::Global
     {
     public:
-        Global(wl_display* display);
+        Global(wl_display* display, std::shared_ptr<DecorationStrategy> strategy);
 
     private:
+        std::shared_ptr<DecorationStrategy> decoration_strategy;
         void bind(wl_resource* new_zxdg_decoration_manager_v1) override;
     };
 
 private:
     void get_toplevel_decoration(wl_resource* id, wl_resource* toplevel) override;
+
     ToplevelsWithDecorations toplevels_with_decorations;
+    std::shared_ptr<DecorationStrategy> decoration_strategy;
 };
 
 class XdgToplevelDecorationV1 : public wayland::XdgToplevelDecorationV1
 {
 public:
-    XdgToplevelDecorationV1(wl_resource* id, mir::frontend::XdgToplevelStable* toplevel);
+    XdgToplevelDecorationV1(
+        wl_resource* id, mir::frontend::XdgToplevelStable* toplevel, std::shared_ptr<DecorationStrategy> strategy);
 
     void set_mode(uint32_t mode) override;
     void unset_mode() override;
 
 private:
+    auto decorations_type_to_protocol_mode(DecorationStrategy::DecorationsType) -> uint32_t;
+    auto protocol_mode_to_decorations_type(uint32_t) -> DecorationStrategy::DecorationsType;
```

Comment:
These could be class methods, they don't need to be instance methods. Also, the names seem verbose
```suggestion
    static auto to_mode(DecorationStrategy::DecorationsType) -> uint32_t;
    static auto to_decorations_type(uint32_t) -> DecorationStrategy::DecorationsType;
```

Reply:
Note that `to_decorations_type` had to be reverted to be a class method since I needed to access `client` from inside it. Would it be better to add a parameter in make it into a static function?